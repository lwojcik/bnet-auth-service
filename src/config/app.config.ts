import { registerAs } from '@nestjs/config';
import { trueStringToBoolean } from '../utils/trueStringToBoolean';
import { APP, APP_INFO, DEFAULTS } from '../common/constants';

const { env } = process;
const defaultValue = DEFAULTS.app;

export const appConfig = registerAs('app', () => ({
  name: APP_INFO.name,
  environment: env[APP.environment],
  host: env[APP.host] || defaultValue.host,
  port: parseInt(env[APP.port], 10) || defaultValue.port,
  enableCors:
    trueStringToBoolean({ value: env[APP.enableCors] }) ||
    defaultValue.enableCors,
  corsOrigin: env[APP.corsOrigin] || defaultValue.corsOrigin,
}));
