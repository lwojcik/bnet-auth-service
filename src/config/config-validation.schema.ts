import * as Joi from 'joi';
import { CronExpression } from '@nestjs/schedule';
import { validateRegionName } from '../common/validators/validateRegionName.validator';

const appSchema = {
  NODE_ENV: Joi.string().default('production'),
  BAS_APP_HOST: Joi.string().default('0.0.0.0'),
  BAS_APP_PORT: Joi.string().default(3000),
  BAS_APP_CORS_ORIGIN: Joi.optional(),
};

const redisSchema = {
  BAS_REDIS_HOST: Joi.string().optional().default('redis'),
  BAS_REDIS_PORT: Joi.string().optional().default('6379'),
  BAS_REDIS_PASSWORD: Joi.any().optional(),
  BAS_REDIS_TTL_SECS: Joi.string().optional().default('2000'),
  BAS_REDIS_DB: Joi.string().optional().default('0'),
};

const battlenetSchema = {
  BAS_BATTLENET_REGION: Joi.string().required().custom(validateRegionName),
  BAS_BATTLENET_CLIENT_ID: Joi.string().required(),
  BAS_BATTLENET_CLIENT_SECRET: Joi.string().required(),
};

const throttleSchema = {
  BAS_THROTTLE_LIMIT: Joi.string().optional().default('300'),
  BAS_THROTTLE_TTL_SECS: Joi.string().optional().default('60'),
};

const authSchema = {
  BAS_AUTH_USERNAME: Joi.string().optional(),
  BAS_AUTH_JWT_SECRET: Joi.string().optional(),
};

const httpsSchema = {
  BAS_HTTPS_KEY_PATH: Joi.string().optional(),
  BAS_HTTPS_CERT_PATH: Joi.string().optional(),
};

const cronSchema = {
  BAS_CRON_PATTERN: Joi.string()
    .optional()
    .default(CronExpression.EVERY_30_MINUTES),
};

export const configValidationSchema = Joi.object({
  ...appSchema,
  ...redisSchema,
  ...battlenetSchema,
  ...throttleSchema,
  ...authSchema,
  ...httpsSchema,
  ...cronSchema,
});
