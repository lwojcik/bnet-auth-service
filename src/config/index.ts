export * from './app.config';
export * from './auth.config';
export * from './battlenet.config';
export * from './endpoints.config';
export * from './redis.config';
export * from './throttle.config';
