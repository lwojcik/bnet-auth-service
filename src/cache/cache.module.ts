import { RedisModule } from '@liaoliaots/nestjs-redis';
import { Module } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { redisConfig } from '../config';
import { REDIS } from '../common/constants';
import { CacheService } from './cache.service';
import { LoggerModule } from '../logger/logger.module';

@Module({
  imports: [
    ConfigModule.forFeature(redisConfig),
    LoggerModule,
    RedisModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: (configService: ConfigService) => ({
        readyLog: true,
        config: {
          host: configService.get(REDIS.host),
          password: configService.get(REDIS.password),
          port: configService.get(REDIS.port),
          db: configService.get(REDIS.db),
          keyPrefix: configService.get(REDIS.keyPrefix),
        },
      }),
    }),
  ],
  providers: [CacheService],
  exports: [CacheService],
})
export class CacheModule {}
