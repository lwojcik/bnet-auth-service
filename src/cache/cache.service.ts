import { RedisService } from '@liaoliaots/nestjs-redis';
import { Inject, Injectable } from '@nestjs/common';
import { ConfigType } from '@nestjs/config';
import Redis from 'ioredis';
import { LoggerService } from '../logger/logger.service';
import { redisConfig } from '../config';
import { REDIS } from '../common/constants';

@Injectable()
export class CacheService {
  private readonly cache: Redis;

  private readonly cacheKey: string;

  constructor(
    private readonly redisService: RedisService,
    @Inject(redisConfig.KEY)
    private readonly redisConf: ConfigType<typeof redisConfig>,
    private readonly logger: LoggerService
  ) {
    this.cache = this.redisService.getClient();
    this.cacheKey = `:${this.redisConf.keyName}`;
    this.logger.setLoggedClass(CacheService.name);
  }

  saveAccessToken(accessToken: string) {
    this.logger.setLoggedMethod(this.saveAccessToken.name, accessToken);
    this.logger.debug();

    if (!this.redisConf.enable) {
      this.logger.debug(`${REDIS.enable} set to false - access key not saved`);
    } else {
      this.logger.debug(
        `Using Redis key: ${this.redisConf.keyPrefix}${this.cacheKey}`
      );

      this.logger.debug(`Set TTL to ${this.redisConf.ttlSecs} seconds`);

      this.cache.set(this.cacheKey, accessToken, 'EX', this.redisConf.ttlSecs);
    }
  }

  async getAccessToken() {
    this.logger.setLoggedMethod(this.getAccessToken.name);
    this.logger.debug();

    if (!this.redisConf.enable) {
      this.logger.debug(
        `${REDIS.enable} set to false - returning 'null' as access key`
      );
      return null;
    }

    this.logger.debug(
      `Using Redis key: ${this.redisConf.keyPrefix}${this.cacheKey}`
    );

    const accessToken = await this.cache.get(this.cacheKey);

    this.logger.debug(`Received access token: ${accessToken}`);

    return accessToken;
  }
}
