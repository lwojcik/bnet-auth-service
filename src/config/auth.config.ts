import { registerAs } from '@nestjs/config';
import { trueStringToBoolean } from '../utils/trueStringToBoolean';
import { AUTH, DEFAULTS } from '../common/constants';

const { env } = process;
const defaultValue = DEFAULTS.auth;

export const authConfig = registerAs('auth', () => ({
  enable:
    trueStringToBoolean({ value: process.env[AUTH.enable] }) ||
    defaultValue.enable,
  username: env[AUTH.username] || defaultValue.username,
  jwtSecret: env[AUTH.jwtSecret] || defaultValue.jwtSecret,
}));
