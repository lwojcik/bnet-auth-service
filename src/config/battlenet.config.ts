import { registerAs } from '@nestjs/config';
import { BATTLENET } from '../common/constants';

const { env } = process;

export const battleNetConfig = registerAs('battlenet', () => ({
  region: env[BATTLENET.region],
  clientId: env[BATTLENET.clientId],
  clientSecret: env[BATTLENET.clientSecret],
}));
