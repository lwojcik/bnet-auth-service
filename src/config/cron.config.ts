import { registerAs } from '@nestjs/config';

export const cronConfig = registerAs('cron', () => ({
  enable: process.env.BAS_CRON_ENABLE === 'true',
  pattern: process.env.BAS_CRON_PATTERN,
}));
