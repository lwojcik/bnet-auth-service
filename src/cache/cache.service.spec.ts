import { RedisService } from '@liaoliaots/nestjs-redis';
import { ConfigModule, registerAs } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { LoggerService } from '../logger/logger.service';
import { CacheService } from './cache.service';

describe('CacheService', () => {
  let service: CacheService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forFeature(
          registerAs('redis', () => ({
            enable: true,
            host: 'test-redis-host',
            port: '1234',
            password: 'test-password',
            ttlSecs: 10,
            db: 0,
            keyName: 'testkeyname',
            keyPrefix: 'testkeyprefix',
          }))
        ),
      ],
      providers: [
        {
          provide: RedisService,
          useValue: {
            getClient: () => ({
              get: jest
                .fn()
                .mockImplementation(
                  () => 'sample_access_token_from_mocked_redis_cache'
                ),
              set: jest.fn(),
            }),
          },
        },
        {
          provide: LoggerService,
          useValue: {
            debug: jest.fn(),
            log: jest.fn(),
            setLoggedClass: jest.fn(),
            setLoggedMethod: jest.fn(),
          },
        },
        CacheService,
      ],
    }).compile();

    service = module.get<CacheService>(CacheService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should get access token', async () => {
    expect.assertions(1);
    const accessToken = await service.getAccessToken();
    expect(accessToken).toMatchSnapshot();
  });
});
