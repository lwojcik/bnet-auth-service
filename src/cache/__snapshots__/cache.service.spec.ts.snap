// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheService should get access token 1`] = `"sample_access_token_from_mocked_redis_cache"`;
